src/compiler/checker.ts

```diff
@@ -7264,6 +7264,16 @@ module ts {
         function checkTypeAssertion(node: TypeAssertion): Type {
             let exprType = checkExpression(node.expression);
             let targetType = getTypeFromTypeNode(node.type);
+            return checkAssertion(node, exprType, targetType);
+        }
+
+        function checkAsExpression(node: AsExpression) {
+            let exprType = checkExpression(node.left);
+            let targetType = getTypeFromTypeNode(node.right);
+            return checkAssertion(node, exprType, targetType);
+        }
+
+        function checkAssertion(node: Node, exprType: Type, targetType: Type) {

```

Comment:
You might consider having both `SyntaxKind.TypeAssertionExpression` and `SyntaxKind.AsExpression` dispatch directly to this method with a `node` parameter of type `TypeAssertionExpression | AsExpression`.